index.js

```diff
@@ -0,0 +1,73 @@
+const express = require('express');
+const bodyParser = require('body-parser');
+const patientGreetingService = require('./services/patient-greeting');
+
+const app = express();
+
+// This is necessary middleware to parse JSON into the incoming request body for POST requests
+app.use(bodyParser.json());
+
+app.use((request, response, next) => {
+  if (request.method !== 'GET' && request.headers && request.headers['content-type']
+    && request.headers['content-type'].match('json\\+fhir')) {
+    request.body = request.body ? JSON.parse(request.body.toString()) : {};
+  }
+
+  next();
+});
+
+// CDS Services must implement CORS to be called from a web browser
+app.use((request, response, next) => {
+  response.setHeader('Access-Control-Allow-Origin', '*');
+  response.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
+  response.setHeader('Access-Control-Allow-Credentials', 'true');
+  response.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
+  response.setHeader('Access-Control-Expose-Headers', 'Origin, Accept, Content-Location, ' +
+    'Location, X-Requested-With');
+
+  next();
+});
+
+// Keys of this object should be CDS Service ID's
+const services = {
+  'patient-greeting': patientGreetingService,
+};
+
+const isValidJson = (obj) => {
+  try {
+    JSON.stringify(obj);
+    return true;
+  } catch (err) {
+    return false;
+  }
+};
+
+// GET: Discovery endpoint
+app.get('/cds-services', (request, response) => {
+  const discoveryEndpointServices = {
+    services: Object.keys(services).map(serviceId => services[serviceId].definition),
+  };
+  response.send(JSON.stringify(discoveryEndpointServices, null, 2));
+});
+
+// POST: Individual CDS Services through multiple serviceId handlers
+app.post('/cds-services/:serviceId', (request, response) => {
+  const serviceEndpoint = request.params.serviceId;
+
+  // Invalid CDS Service endpoint
+  if (!(serviceEndpoint in services)) {
+    response.status(404);
+  }
+
+  // Bad request
+  if (!isValidJson(request.body)) {
+    response.status(400);
+  }
+
+  const cdsService = services[serviceEndpoint];
```

Comment:
In either of the above error scenarios (404 or 500), we should be returning immediately. Instead, execution flow continues to this `services` map lookup.

For instance, run the server locally and do a POST to http://localhost:3000/cds-services/foo. You'll see a log error of:

```
TypeError: Cannot read property 'serviceResponse' of undefined
    at app.post (/Users/shek/dev/sandbox-cds-services/index.js:68:13)
```